File: circularDobList.py
import threading
import time

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class CircularDoublyLinkedList:
    def __init__(self):
        self.head = None
        self.size = 0
        self.isRunning = False

    # Tambah Berita
    def tambah_berita(self, teks):
        new_node = Node(teks)

        if self.head is None:
            self.head = new_node
            self.head.next = self.head
            self.head.prev = self.head
        else:
            tail = self.head.prev
            tail.next = new_node
            new_node.prev = tail
            new_node.next = self.head
            self.head.prev = new_node

        self.size += 1
        print(f"Data Berita: {new_node.data} berhasil ditambahkan!")

    # Stop running
    def stop_running(self):
        self.isRunning = False
        print("\n--- Tampilan Berita Berhenti ---")

    # Forward
    def forward_berita(self):
        if self.head is None:
            return

        self.isRunning = True
        curr = self.head

        while self.isRunning:
            print(f"<<< {curr.data} >>>")
            time.sleep(3)
            curr = curr.next

    # Backward
    def backward_berita(self):
        if self.head is None:
            return

        self.isRunning = True
        curr = self.head.prev

        while self.isRunning:
            print(f"<<< {curr.data} >>>")
            time.sleep(3)
            curr = curr.prev

    # Hapus
    def hapus_berita(self, pos):
        if self.head is None or pos < 1 or pos > self.size:
            print("Gagal Menghapus: posisi tidak valid")
            return

        curr = self.head
        for _ in range(1, pos):
            curr = curr.next

        if self.size == 1:
            self.head = None
        else:
            curr.prev.next = curr.next
            curr.next.prev = curr.prev

            if curr == self.head:
                self.head = curr.next

        self.size -= 1
        print(f"Data Berita: {curr.data} berhasil dihapus!")

    # Tampilkan berdasarkan posisi
    def tampilkan_berita(self, pos):
        if self.head is None:
            print("Berita kosong!")
            return

        if pos < 1 or pos > self.size:
            print("Nomor tidak ditemukan!")
            return

        curr = self.head
        for _ in range(1, pos):
            curr = curr.next

        print(f"Data: {curr.data}")


# ================= MAIN =================
def main():
    list_berita = CircularDoublyLinkedList()

    while True:
        print("\n==== Sistem Running Text Berita ====")
        print("1. Tambah Berita")
        print("2. Hapus Berita")
        print("3. Running Text Forward")
        print("4. Running Text Backward")
        print("5. Tampilkan Berita")
        print("6. Keluar")

        pilihan = int(input("Pilih: "))

        if pilihan == 1:
            teks = input("Masukkan berita: ")
            list_berita.tambah_berita(teks)

        elif pilihan == 2:
            pos = int(input("Masukkan nomor berita: "))
            list_berita.hapus_berita(pos)

        elif pilihan == 3:
            print("(Tekan ENTER untuk berhenti...)")

            t = threading.Thread(target=list_berita.forward_berita)
            t.start()

            input()  # tunggu ENTER
            list_berita.stop_running()
            t.join()

        elif pilihan == 4:
            print("(Tekan ENTER untuk berhenti...)")

            t = threading.Thread(target=list_berita.backward_berita)
            t.start()

            input()
            list_berita.stop_running()
            t.join()

        elif pilihan == 5:
            pos = int(input("Masukkan nomor berita: "))
            list_berita.tampilkan_berita(pos)

        elif pilihan == 6:
            print("Program selesai")
            break


if __name__ == "__main__":
    main()

File: circularDobList.java

import java.util.Scanner;

class Node {
    String data;
    Node next, prev;

    public Node(String data) {
        this.data = data;
    }

}

class CircularDoublyLinkedList{
    Node head = null;
    int size = 0;
// Tambah Berita
    void tambahBerita(String teks){
        Node newNode = new Node(teks);
        if (head == null) {
            head = newNode;
            head.next = head;
            head.prev = head;
        } else {
            Node tail = head.prev;
            tail.next = newNode;
            newNode.prev = tail;
            newNode.next = head;
            head.prev = newNode;
        }
        size++;
        System.out.println("Data Berita: "+ newNode.data + " Berhasil Ditambahkan!");
    }

    public volatile boolean isRunning = false;

    public void stopRunning(){
        isRunning = false;
        System.out.println("---Tampilan Berita Berhenti---");
    }
// Forward Tampilan Berita
    void forwardBerita() throws InterruptedException{
        if (head == null) {
            return;
        }
        isRunning = true;
        Node currNode  = head;
        while(isRunning){
            System.out.println("<<< " + currNode.data + " >>>");

            try{
                Thread.sleep(3000);
            } catch (InterruptedException e){
                System.out.println("Sistem Intrupsi");
                break;
            }

            currNode = currNode.next;
        }
    }
// Backward Tampilan Berita
    void  backwardBerita() throws InterruptedException{
        if (head == null) {
            return;
        }
        isRunning = true;
        Node currNode = head.prev;

        while (isRunning) {
            System.out.println("<<< " + currNode.data + " >>>");

            try{
                Thread.sleep(3000);
            } catch(InterruptedException e){
                System.out.println("Sistem Intrupsi");
                break;
            }
            
            currNode = currNode.prev;
        }
    }
// Hapus Berita
    void hapusBerita(int pos){
        if (head == null || pos < 1 || pos > size) {
            System.out.println("Gagal Menghapus Berita: Posisi Input Tidak Valid");
            return;
        }
        Node curr = head;
        for(int i = 1; i < pos; i++){
            curr = curr.next;
        }

        if (size == 1) {
            head = null;
        } else {
            curr.prev.next = curr.next;
            curr.next.prev = curr.prev;
            if (curr == head) {
                head = curr.next;
            }
        }
        size--;

        System.out.println("Data Berita: " + curr.data + " Berhasil Dihapus!");
    }
// Tampilkan Berita
    void tampilanPilihanBerita(int pos){
        if (head == null) {
            System.out.println("Berita Kosong!");
            return;
        }
        if (pos > size || pos < 1) {
            System.out.println("Nomor Berita Tidak Ditemukan!");
            return;
        }
        Node curr = head;
        for(int i = 1; i < pos; i++){
            curr = curr.next;
        }
        System.out.println("Data: "+ curr.data);

    }
}

public class circularDobList {
    public static void main(String[] args) throws InterruptedException {
            Scanner sc = new Scanner(System.in);
            int pilihanMenu;
            CircularDoublyLinkedList list = new  CircularDoublyLinkedList();
            System.out.println("==== Sistem Running Text Berita ====");
            do { System.out.println("=============================================");
                System.out.println("Pilih Menu: ");
                System.out.println("1. Tambahkan Data Berita ");
                System.out.println("2. Hapus Data Berita ");
                System.out.println("3. Tampilkan Running Text Berita Forward (3s) ");
                System.out.println("4. Tampilkan Runing Text Berita Backward (3s) ");
                System.out.println("5. Tampilkan Data Berita");
                System.out.println("6. Keluar");
                System.out.println("=============================================");
                System.out.println("Pilihan: ");

                pilihanMenu = sc.nextInt();
                sc.nextLine();
                switch (pilihanMenu) {
                    case 1:
                        System.out.println("=== Tambah Berita ===");
                        System.out.println("Masukkan Text Berita: ");
                        String dataBerita = sc.nextLine();

                        list.tambahBerita(dataBerita);
                        break;
                    case 2: 
                        System.out.println("=== Hapus Berita ===");
                        System.out.println("Masukkan Nomor Berita: ");
                        list.hapusBerita(sc.nextInt());
                        break;
                    case 3:
                        System.out.println("=== Tampil Forward Berita ===");
                        System.out.println("(Tekan tombol ENTER kapan saja untuk berhenti...)");
                        System.out.println("----------------------------------------------");

                        Thread threadBerita = new Thread(() -> {
                            try {
                                list.forwardBerita();
                            } catch (InterruptedException e) {
                            }
                        });
                        threadBerita.start();
                        sc.nextLine(); 

                        sc.nextLine(); 

                        list.isRunning = false; 
                        threadBerita.interrupt();

                        System.out.println("\n=============================================");
                        System.out.println("Berita berhasil dihentikan. Kembali ke menu utama.");
                        break;
                    case 4:
                        System.out.println("=== Tampil Backward Berita ===");
                        System.out.println("(Tekan tombol ENTER kapan saja untuk berhenti...)");
                        System.out.println("----------------------------------------------");

                        Thread threadBeritaBack = new Thread(() -> {
                            try {
                                list.backwardBerita();
                            } catch (InterruptedException e) {
                            }
                        });

                        threadBeritaBack.start();

                        sc.nextLine(); 
                        sc.nextLine(); 

                        list.isRunning = false;
                        threadBeritaBack.interrupt();

                        System.out.println("\n=============================================");
                        System.out.println("Berita berhasil dihentikan. Kembali ke menu utama.");
                        break;
                    case 5:
                        System.out.println("=== Cari Berita ===");
                        System.out.println("Masukkan Nomor Berita: ");
                        int pos = sc.nextInt();
                        list.tampilanPilihanBerita(pos);
                        break;
                    case 6:
                        sc.close();
                        System.out.println("Program Telah Ditutup");
                        System.exit(0);
                        break;
                }
            } while (true);
    }
}